scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,824 @@
+            throws InvalidChannelException, CompilationException, IOException {
+        PredicateActionResult<T> answer;
+        try {
+            Channel<T> chosenChannel = getChannel(channel);
+            T value = chosenChannel.get(key);
+            if (evaluateClosure(predicate, BiPredicate.class).test(key, value)) {
+                answer = new PredicateActionResult<>(true,
+                                                     chosenChannel.compute(key,
+                                                                           (BiFunction<? super String, ? super T, ? extends T>) evaluateClosure(thenRemappingFunction,
+                                                                                                                                                BiFunction.class)));
+            } else {
+                throw new IllegalStateException("Predicate " + predicate + " is not met");
+            }
+            commitIfNeeded(channel);
+            logWithContext(taskid, channel, "Run waitUntil on key '" + key + "' result is: " + answer, Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new ClosureEvaluationException("Exception when evaluating closure : " +
+                                                 StackTraceUtil.getStackTrace(e));
+        }
+        return answer;
+    }
+
+    private void logWithContext(TaskId taskId, String channel, String message, Level level)
+            throws InvalidChannelException {
+        String channelHeader = "[Synchronization]" + (channel != null ? "[Channel=" + channel + "] " : "");
+        switch (level.toInt()) {
+            case Level.TRACE_INT:
+                if (logger.isTraceEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.DEBUG_INT:
+                if (logger.isDebugEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.INFO_INT:
+                if (logger.isInfoEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.WARN_INT:
+                TaskLogger.getInstance().warn(taskId, channelHeader + message);
+                break;
+            case Level.ERROR_INT:
+                TaskLogger.getInstance().error(taskId, channelHeader + message);
+                break;
+            default:
+                // no action
+        }
+        if (logger.isTraceEnabled() && channel != null) {
+            TaskLogger.getInstance().info(taskId, channelHeader + "New channel content : " + getChannel(channel));
+        }
+    }
+
+    @Override
+    public void runActivity(Body body) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)